src/Staged/GHC/Generics/GHCish.hs

```diff
@@ -0,0 +1,149 @@
+{-# language AllowAmbiguousTypes #-}
+{-# language DataKinds #-}
+{-# language EmptyCase #-}
+{-# language FlexibleContexts #-}
+{-# language FlexibleInstances #-}
+{-# language KindSignatures #-}
+{-# language PolyKinds #-}
+{-# language ScopedTypeVariables #-}
+{-# language StandaloneKindSignatures #-}
+{-# language TemplateHaskellQuotes #-}
+{-# language TypeApplications #-}
+{-# language TypeFamilies #-}
+{-# language TypeOperators #-}
+
+module Staged.GHC.Generics.GHCish
+  ( ghcGenericTo
+  , ghcGenericFrom
+  ) where
+
+import Staged.GHC.Generics.Types
+import qualified GHC.Generics as G
+import Language.Haskell.TH.Syntax hiding (Type)
+import Language.Haskell.TH.Lib
+import Data.Kind (Type, Constraint)
+import Data.Bifunctor (first)
+
+-- A proxy type for GHC.Generics functions
+data Prox (d :: Meta) (f :: j -> Type) (a :: j) = Prox
+toProx :: forall j (q :: Type -> Type) i (d :: Meta) (f :: (Type -> Type) -> j -> Type) (a :: j).
+  M2 i d f q a -> Prox d (f q) a
+toProx _ = Prox
+
+-- | Use GHC generics to implement the 'to' method.
+ghcGenericTo :: (Rep a ~ Translate (G.Rep a), G.Generic a, GGeneric (Rep a), Quote q)
+  => Rep a (Code q) x -> Code q a
+ghcGenericTo = unsafeCodeCoerce . gto
+
+{-
+-- `from` is not the most obvious thing, in general. I found it quite
+-- helpful to look at an example of what we're trying to build.
+
+data MyType a = My1 a | My2 a a
+  deriving (G.Generic)
+
+instance Generic (MyType a) where
+  from x k = unsafeCodeCoerce $ caseE (unTypeCode x) [
+      do
+        v1 <- newName "v1"
+        match (conP 'My1 [varP v1])
+              (normalB $ unTypeCode $ k (M2 (L2 (M2 (M2 (K2 (unsafeCodeCoerce $ varE v1)))))))
+              []
+    , do
+        v1 <- newName "v1"
+        v2 <- newName "v2"
+        match (conP 'My2 [varP v1, varP v2])
+              (normalB $ unTypeCode $ k (M2 (R2 (M2
+                (M2 (K2 (unsafeCodeCoerce $ varE v1)) :**: M2 (K2 (unsafeCodeCoerce $ varE v2)))))))
+              []
+    ]
+-}
+
+-- | Use GHC generics to implement the 'from' method.
+ghcGenericFrom :: (Rep a ~ Translate (G.Rep a), G.Generic a, GGeneric (Rep a), Quote q)
+  => Code q a -> (Rep a (Code q) x -> Code q r) -> Code q r
+ghcGenericFrom c k = unsafeCodeCoerce $ caseE (unTypeCode c) $ gmatches k
+
+class GGeneric (f :: (Type -> Type) -> Type -> Type) where
+  gto :: Quote q => f (Code q) x -> q Exp
+
+  -- Build the list of case branches.
+  gmatches :: Quote q => (f (Code q) x -> Code q r) -> [q Match]
```

Comment:
I use four space indentation. I'll reformat this if you won't.

Reply:
I'm a bit embarrassed to admit I don't know how to do that (other than manually). What tool do you use?

Reply:
I do tricks with VIM `:retab` or do it manually, depends on the day :)

Reply:
Couldn't figure that out as yet. If you can do it, I'd appreciate that. I think this is ready for another review.